### PR TITLE
Add support for the Bosch BME680 Environmental Sensor

### PR DESCRIPTION
**Pull request by amrbekhit**
This adds support for the BME680 sensor to the mynewt sensor framework.
The driver supports both I2C and SPI modes (NOTE: SPI not tested at time
of commit). The syscfg file can be used to configure whether or not
floating point compensation is used or not, and whether the output values
use the standard mynewt sensor value types (found in temperature.h,
pressure.h and humidity.h) or the Bosch definitions (using struct bme680_field_data
in bme680_defs.h). The gas resistance output is either a float or a uin32_t,
depending on whether floating point compensation is turned on or not.

The sensor can be used standalone with this driver, but is ultimately
designed to be used alongside the Bosch BSEC sensor fusion library for
best performance (https://www.bosch-sensortec.com/bst/products/all_products/bsec).
A mynewt module adding support for this library will be added at a later date.

The sensor_creator code has been modified to include an off-board instance of the
sensor using I2C0. This can be enabled usign the BME680_OFB syscfg value.

### CODE REVIEW

File: hw/drivers/sensors/bme680/src/bme680.c [hunk 1]
```diff
@@ -0,0 +1,1694 @@
...
+static int8_t boundary_check(uint8_t *value, uint8_t min, uint8_t max, struct bme680_cfg *dev);
+
+/*!
+ *@brief This API is the entry point.
+ *It reads the chip-id and calibration data from the sensor.
+ */
+int8_t 
+bme680_internal_init(struct bme680_cfg *dev)
+{
+	int8_t rslt;
```
Comment (sterlinghughes) [inline hw/drivers/sensors/bme680/src/bme680.c]: coding standards -- please use 4 spaces versus tabs or 8 spaces.

File: hw/drivers/sensors/bme680/src/bme680.c [hunk 2]
```diff
@@ -0,0 +1,1694 @@
...
+	rslt = null_ptr_check(dev);
+	if (rslt == BME680_OK) {
+		if (dev->intf == BME680_SPI_INTF) {
+			/* Set the memory page */
+			rslt = set_mem_page(reg_addr, dev);
+			if (rslt == BME680_OK)
+				reg_addr = reg_addr | BME680_SPI_RD_MSK;
+		}
+		dev->com_rslt = dev->read(dev->sensor, dev->dev_id, reg_addr, reg_data, len);
+		if (dev->com_rslt != 0)
```
Comment (sterlinghughes) [inline hw/drivers/sensors/bme680/src/bme680.c]: coding standards: all statements must be enclosed in braces, e.g.  no single line if's

File: hw/drivers/sensors/bme680/src/bme680.c [hunk 3]
```diff
@@ -0,0 +1,1694 @@
...
+{
+	int8_t rslt;
+
+	/* Check for null pointer in the device structure*/
+	rslt = null_ptr_check(dev);
+	if (rslt == BME680_OK) {
+		if (dev->intf == BME680_SPI_INTF) {
+			/* Set the memory page */
+			rslt = set_mem_page(reg_addr, dev);
+			if (rslt == BME680_OK)
```
Comment (sterlinghughes) [inline hw/drivers/sensors/bme680/src/bme680.c]: no single line if

File: hw/drivers/sensors/bme680/src/bme680.c [hunk 4]
```diff
@@ -0,0 +1,1694 @@
...
+					tmp_buff[(2 * index)] = reg_addr[index] & BME680_SPI_WR_MSK;
+				} else {
+					tmp_buff[(2 * index)] = reg_addr[index];
+				}
+				tmp_buff[(2 * index) + 1] = reg_data[index];
+			}
+			/* Write the interleaved array */
+			if (rslt == BME680_OK) {
+				dev->com_rslt = dev->write(dev->sensor, dev->dev_id, tmp_buff[0], &tmp_buff[1], (2 * len) - 1);
+				if (dev->com_rslt != 0)
```
Comment (sterlinghughes) [inline hw/drivers/sensors/bme680/src/bme680.c]: single line if

File: hw/drivers/sensors/bme680/src/bme680.c [hunk 5]
```diff
@@ -0,0 +1,1694 @@
...
+	if (rslt == BME680_OK) {
+		if (reg_addr > 0x7f)
+			mem_page = BME680_MEM_PAGE1;
+		else
+			mem_page = BME680_MEM_PAGE0;
+
+		if (mem_page != dev->mem_page) {
+			dev->mem_page = mem_page;
+
+			dev->com_rslt = dev->read(dev->sensor, dev->dev_id, BME680_MEM_PAGE_ADDR | BME680_SPI_RD_MSK, &reg, 1);
```
Comment (sterlinghughes) [inline hw/drivers/sensors/bme680/src/bme680.c]: line length must be less than 80 chars.  github doesn't tell me line length: but I'd bet this is over :)

File: hw/drivers/sensors/bme680/src/bme680.c [hunk 6]
```diff
@@ -0,0 +1,1694 @@
...
+
+static int8_t
+bme680_i2c_write(struct sensor *sensor, uint8_t dev_id, uint8_t reg_addr, uint8_t *reg_data, uint16_t len)
+{
+	int8_t rc = 0;
+	struct sensor_itf *interface = SENSOR_GET_ITF(sensor);
+	struct hal_i2c_master_data data;
+	uint8_t *buf;
+
+	buf = malloc(len + 1);
```
Comment (sterlinghughes) [inline hw/drivers/sensors/bme680/src/bme680.c]: avoid malloc() and free() in running code -- how big are these lengths generally?  better: use a buffer on the stack if not too big, or malloc() a single buffer at init time.

**Comment by amrbekhit** [comment]
@sterlinghughes Thanks for the review - I'm going through and making the changes. When writing this, I didn't touch the third party code except when necessary, but applied the coding styles to my own additions. Should the third party code be modified to confirm to the coding standards (e.g. regarding things like having function return types be on a separate line)? Some third party files (such as stm32_driver_mod_i2c_v2.c) don't seem to do this.

**Comment by sterlinghughes** [comment]
hi @amrbekhit -- thanks for updating, if 3rd party code is in a separate file no need.  If you've copied it into a newly written file, please update that to coding standards.

**Comment by amrbekhit** [comment]
@sterlinghughes I've push another commit. Hopefully this should clean up the style and also eliminate the use of malloc/free. I've also removed an unused header file and made all global functions that don't need to be global static.

**Comment by amrbekhit** [comment]
The pointers are initialised in bme680_config so that they point to either
the spi or i2c functions. The driver then calls these without worrying
about the interface.

File: hw/drivers/sensors/bme680/include/bme680/bme680.h [hunk 7]
```diff
@@ -0,0 +1,26 @@
+#ifndef __BME860_H__
```
Comment (vrahane) [inline hw/drivers/sensors/bme680/include/bme680/bme680.h]: Needs Apache copyright header.

File: hw/drivers/sensors/bme680/include/bme680/bme680_defs.h [hunk 8]
```diff
@@ -0,0 +1,557 @@
...
+ * of such information nor for any infringement of patents or
+ * other rights of third parties which may result from its use.
+ * No license is granted by implication or otherwise under any patent or
+ * patent rights of the copyright holder.
+ *
+ * @file	bme680_defs.h
+ * @date	22 Feb 2018
+ * @version	3.5.8
+ * @brief
+ *
```
Comment (vrahane) [inline hw/drivers/sensors/bme680/include/bme680/bme680_defs.h]: Needs Apache copyright header.

File: hw/drivers/sensors/bme680/src/bme680.c [hunk 9]
```diff
@@ -0,0 +1,2030 @@
...
+ * other rights of third parties which may result from its use.
+ * No license is granted by implication or otherwise under any patent or
+ * patent rights of the copyright holder.
+ *
+ * File     bme680.c
+ * @date    22 Feb 2018
+ * @version 3.5.8
+ *
+ */
+
```
Comment (vrahane) [inline hw/drivers/sensors/bme680/src/bme680.c]: Needs Apache copyright header.

File: hw/drivers/sensors/bme680/pkg.yml [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+#
+
+pkg.name: hw/drivers/sensors/bme680
+pkg.description: Driver for the Bosch BME680 environmental sensor
+pkg.author: "Amr Bekhit"
```
Comment (vrahane) [inline hw/drivers/sensors/bme680/pkg.yml]: I know some of the sensors have the author as the authors name but this needs to be "Apache Mynewt".

**Comment by vrahane** [comment]
Sorry, realized after merging the PR but there is an issue with the copyright, as most of the sensors, all of them somehow got copied from the same copyright header which was fixed recently. In all the files can you please replace the following text. Change "resarding" to "regarding".

**Comment by amrbekhit** [comment]
@vrahane Done - new PR at https://github.com/apache/mynewt-core/pull/1487

**Pull request merged**